pkg/record/record.go

```diff
@@ -0,0 +1,71 @@
+package record
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"io"
+)
+
+// ErrIllegalTopicName is returned if a topic's character sequence is invalid.
+var ErrIllegalTopicName = errors.New("illegal topic name")
+
+// Reader emits records.
+// It returns io.EOF if the underlying record source has no more data.
+type Reader func() (record []byte, err error)
```

Comment:
I'm a bit late to the party, but I'd like to know if we considered to reuse the `io.Reader` interface for the same effect, instead of defining a new one. This all seems similar to what RecordIO does: https://github.com/mesos/mesos-go/blob/master/api/v1/lib/recordio/doc.go

```go



```

Reply:
The only difference to an `io.Reader` is essentially where the `[]byte` is placed. If we made it an argument, the caller had to pass in a sufficiently large slice and than truncate it based on the returned number of bytes read.  
This will make it more complicated again for the caller – especially for allowing arbitrary record sizes. Though actually "arbitrary" probably isn't something we want?

OTOH, if we ever conclude that we want to optimize allocs, it might be easier if the caller can pass in their own `[]byte`s. There are probably other ways though.

The other reservation I have is that `io.Reader` comes with certain set of assumptions of how it can be used (and, for example, fed into various things like `bufio.Reader`s) that don't align well with the fixed record semantics we actually have.